cheroot/ssl/builtin.py

```diff
@@ -277,6 +277,11 @@ def wrap(self, sock):
             s = self.context.wrap_socket(
                 sock, do_handshake_on_connect=True, server_side=True,
             )
+        except ssl.SSLZeroReturnError:
+            # This is almost certainly due to the cherrypy engine
+            # 'pinging' the socket to assert it's connectable;
+            # the 'ping' isn't SSL.
+            return EMPTY_RESULT
         except ssl.SSLError as ex:
             if ex.errno == ssl.SSL_ERROR_EOF:
```

Comment:
I wonder if we could declare an internal `CherootSSLZeroReturnError` in `cheroot._compat`, and then wrap the TLS calls to re-raise both `ssl.SSLZeroReturnError` and `ssl.SSLError` with `ssl.SSL_ERROR_EOF` as this internal exception. This would allow having a cleaner structural block for handling two incarnations of the same situation as one thing without having to care about two copy-pasted blocks are handled the same way.

Reply:
Oh, I see, we already have `errors.NoSSLError` so maybe not `_compat` but `errors` would be a better place for the exception.

Reply:
And maybe all these branchy cases could be treated the same with a shorter and better readable error handling on this layer of abstraction.

Reply:
in doing a full review of how this exception is handled, and comparing to the pyopenssl implementation, I would say my opinion is that we should change this entire exception code to just swallow the exception and return EMPTY_RESULT, except for the   http-on-https port condition.  

Something like this.
```
try:
            s = self.context.wrap_socket(
                sock, do_handshake_on_connect=True, server_side=True,
            )
            return s, self.get_environ(s)
except ssl.SSLError as ex:
            if ex.errno == ssl.SSL_ERROR_SSL:
                     if _assert_ssl_exc_contains(ex, 'http request'):
                           # The client is speaking HTTP to an HTTPS server.
                           raise errors.NoSSLError
except generic_socket_error:
           pass
except  ssl.SSLZeroReturnError:
           pass
return EMPTY_RESULT
```
If we ever wanted to log these bad sockets, we can add to this, but right now we dissect a bunch of exceptions all for it to be ignored anyway.

I would also get rid of the `except Exception:` in `server.serve()` as there shouldn't be any unknown/unexpected exceptions bubbling up there, the only unknown/unexpected exceptions should be coming from user code, and the workers are in different threads.

Reply:
My understanding is that the reason this is more complicated than pyopenssl is that pyopenssl does some processing internally and turns weird cases into specific exceptions. I'd love the same to happen with the `ssl` implementation. But it'd have to happen on our side since stdlib doesn't do this. And even if it did, we'd need to have something for compatibility for years.

> I would also get rid of the `except Exception:` in `server.serve()` as there shouldn't be any unknown/unexpected exceptions bubbling up there, the only unknown/unexpected exceptions should be coming from user code, and the workers are in different threads.

Agreed. Although, this deserves a separate discussion/PR.